### PR TITLE
[2.7] bpo-33021: Release the GIL during fstat() calls (GH-6019)

### PR DESCRIPTION
**Pull request by nirs**
If the file descriptor is on a non-responsive NFS server, calling
fstat() can block for long time, hanging all threads. Most of the calls
release the GIL around the call, but some calls seems to be forgotten.

This patch fixes the last calls, affecting users of:
- imp.load_dynamic()
- imp.load_source()
- mmap.mmap()
- mmapobject.size()
- os.fdopen()
- os.urandom()
- random.seed()

https://bugs.python.org/issue33021

### CODE REVIEW

**Review by pitrou — CHANGES_REQUESTED** [review]
As on the other PR, this looks good on the principle, but you'll need to generate a NEWS entry.

Also, since 2.7 is in maintenance mode, I'm cc'ing our release manager Benjamin to get his approval on this. @benjaminp

**Comment by nirs** [comment]
Thanks for the review @pitrou! I have made the requested changes; please review again.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks for the update. This looks good on the principle. I'll defer to @benjaminp for the backport decision.

**Comment by benjaminp** [comment]
Releasing the GIL in new places tends to open new and exciting opportunities for reëntrancy bugs, so I'm not sure this is completely safe.

**Comment by pitrou** [comment]
I'm not sure about reëntrancy bugs, but I agree there may be thread-safetÿ issues on ill-behaving platforms (though for fstat() we already release the GIL in other places).

**Comment by pitrou** [comment]
Actually, you're right, I remember about someone implementing a FUSE filesystem with Python and there were interesting reëntrancy issues with GIL releasing specifics.

**Comment by pitrou** [comment]
Ok, I'm closing this PR as rejected.

**Comment by nirs** [comment]
@benjaminp, we are releasing the GIL on all fstat calls in python 3. Why is this correct and safe in python 3 but no in python 2?

I'm pretty sure that not releasing the GIL is not safe. Try to mmap a file on non-responsive NFS server - do you really want the entire process to hang?

I can extract the mmap.xxx and os.fdopen changes to a separate patch, since the other changes will are much less likely to block (e.g. fstat for /dev/urandom).

**Comment by benjaminp** [comment]
I'm not sure throwing `Py_BEGIN_ALLOW_THREADS` around every `fstat` call is any better for Python 3. Releasing the GIL around operations on `/dev/urandom` FDs is probably pure overhead, since no actual IO ever ends up happening.

**Comment by nirs** [comment]
@benjaminp I suggest to continue on the issue, I added more info there.

**Pull request closed without merging**
